hotrod-transactions/src/main/java/org/jboss/as/quickstarts/datagrid/hotrod/FootballManager.java

```diff
@@ -0,0 +1,245 @@
+   private static final String initialPrompt = "Choose action:\n" + "============= \n"
+         + "at  -  add a team\n"
+         + "ap  -  add a player to a team\n"
+         + "rt  -  remove a team\n"
+         + "rp  -  remove a player from a team\n"
+         + "p   -  print all teams and players\n"
+         + "pc  -  print all players and countries\n"
+         + "q   -  quit\n";
+
+   private RemoteCacheManager cacheManager;
+   private RemoteCache<String, Team> teams;
+   private RemoteCache<String, String> players;
+
+   private static final String TEAMS_CACHE = "teams_tx";
+   private static final String PLAYERS_CACHE = "players_tx";
+
+   private static final String TEAMS_XML_CACHE_CONFIG =
+         "<infinispan><cache-container>" +
+               "  <local-cache-configuration name=\"" + TEAMS_CACHE + "\">" +
+               "    <locking isolation=\"REPEATABLE_READ\"/>" +
+               "    <transaction locking=\"PESSIMISTIC\" mode=\"NON_XA\" />" +
+               "  </local-cache-configuration>" +
+               "</cache-container></infinispan>";
+
+   private static final String PLAYERS_XML_CACHE_CONFIG =
+         "<infinispan><cache-container>" +
+               "  <local-cache-configuration name=\"" + PLAYERS_CACHE + "\">" +
+               "    <locking isolation=\"REPEATABLE_READ\"/>" +
+               "    <transaction locking=\"PESSIMISTIC\" mode=\"NON_XA\" />" +
+               "  </local-cache-configuration>" +
+               "</cache-container></infinispan>";
+
+   private TransactionManager tm;
+
+   private Console con = System.console();
+
+   public FootballManager() {
+      initCaches();
+      // Both caches are using the same tx manager so we can keep the reference
+      tm = teams.getTransactionManager();
+   }
+
+   private void initCaches() {
+      ConfigurationBuilder builder = new ConfigurationBuilder();
+      builder.addServer()
+            .host(jdgProperty(JDG_HOST))
+            .port(Integer.parseInt(jdgProperty(HOTROD_PORT)));
+      // Configure the RemoteCacheManager to use a transactional cache as default
+      // Use the simple TransactionManager in hot rod client
+      builder.transaction().transactionManagerLookup(RemoteTransactionManagerLookup.getInstance());
+      // The cache will be enlisted as Synchronization
+      builder.transaction().transactionMode(TransactionMode.NON_XA);
+
+      cacheManager = new RemoteCacheManager(builder.build());
+      // Create teams as a transactional cache
+      teams = cacheManager.administration().getOrCreateCache(TEAMS_CACHE, new XMLStringConfiguration(TEAMS_XML_CACHE_CONFIG));
+      players = cacheManager.administration().getOrCreateCache(PLAYERS_CACHE, new XMLStringConfiguration(PLAYERS_XML_CACHE_CONFIG));
+   }
+
+   public void addTeam() {
+      String teamName = con.readLine(msgEnterTeamName);
+
+      try {
+         tm.begin();
+         teams.putIfAbsent(teamName, new Team(teamName));
+         tm.commit();
```

Comment:
But I have to handle exceptions again in the finally bloc...

Reply:
if `putIfAbsent()` throws an exception, the `tm.commit()` won't be invoked ever... The transaction will be open forever... the next `tm.begin()` call will fail as well since there is a transaction in progress.

Reply:
Ok, I thought the transaction will rollback. If put if absent fails, what do you think about rollback in the exception ? 
And what happens if commit or rollback fail ?

Reply:
an exception in the client code doesn't change the transaction status. The user is free to do what is best for its use case. You can retry the putIfAbsent, you can ignore the exception or you can rollback the transaction. Only implicit transactions are handled in Infinispan's code

commit() or rollback() can throw exceptions. in your demo, they can be ignored and print an error message.